### PR TITLE
(PXP-7717) Add support for bulk policy endpoints

### PR DESCRIPTION
**Pull request by BinamB**
Jira Ticket: [PXP-xxxx](https://ctds-planx.atlassian.net/browse/PXP-xxxx)
- [ ] Remove this line if you've changed the title to (PXP-xxxx): <title>

### New Features
- Add support for new `/bulk/policy` endpoint
- Add support for new `/user/{username}/bulk/policy` endpoint

### CODE REVIEW

File: python/src/gen3authz/client/arborist/base.py [hunk 1]
```diff
@@ -622,6 +623,23 @@ async def update_policy(self, policy_id, policy_json, create_if_not_exist=False)
...
+        except ArboristError as e:
+            if e.code == 405:
+                # For compatibility with Arborist 2.x.x
+                self.logger.info(
+                    "This Arborist version has no PUT /policy/{policyID} endpt yet."
+                    "Falling back on PUT /policy"
+                )
+                response = await self.put(self._policy_url, json=policy_json)
+            else:
+                raise
```
Comment (paulineribeyre) [inline python/src/gen3authz/client/arborist/base.py:640]: i don't think the try/except is necessary.
if you do want backwards compatibility, you'll have to make multiple calls for each policy in `policy_json`, right?

File: python/src/gen3authz/client/arborist/base.py [hunk 2]
```diff
@@ -622,6 +623,23 @@ async def update_policy(self, policy_id, policy_json, create_if_not_exist=False)
         self.logger.info("put policy {}".format(policy_id))
         return response
 
+    @maybe_sync
+    async def update_bulk_policy(self, policy_json, create_if_not_exist=False):
```
Comment (paulineribeyre) [inline python/src/gen3authz/client/arborist/base.py:627]: `create_if_not_exist` is not used

File: python/src/gen3authz/client/arborist/base.py [hunk 3]
```diff
@@ -686,6 +704,28 @@ async def revoke_user_policy(self, username, policy_id):
...
+        response = await self.post(url, json=request, expect_json=False)
+        if response.code != 204:
+            self.logger.error(
+                "could not grant policies `{}` to user `{}`: {}".format(
+                    policy_ids, username, response.error_msg
+                )
+            )
+            return None
+        self.logger.info(
+            "granted policy `{}` to user `{}`".format(policy_ids, username)
```
Comment (paulineribeyre) [inline python/src/gen3authz/client/arborist/base.py:725]:
```suggestion
            "granted policies `{}` to user `{}`".format(policy_ids, username)
```

File: python/src/gen3authz/client/arborist/base.py (hunk 1 above)
Comment (BinamB) [reply python/src/gen3authz/client/arborist/base.py:640]: Yeah it doesnt evenn matter if its backwards compatible. This is a completely different function. I just copy pasted and forgot to clean up :grimacing:

File: python/src/gen3authz/client/base.py [hunk 4]
```diff
@@ -67,6 +67,10 @@ def get_policy(self, policy_id):
     def update_policy(self, policy_id, policy_json):
         pass
 
+    @abc.abstractmethod
+    def update_bulk_policy(self, policy_json):
```
Comment (paulineribeyre) [inline python/src/gen3authz/client/base.py:71]: do you need to add `grant_bulk_user_policy` in here too?

**Review by paulineribeyre — APPROVED** [review]

**Pull request merged**